ibis/common/graph.py

```diff
@@ -0,0 +1,157 @@
+"""
+Various traversal utilities for the expression graph.
+"""
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections import deque
+from typing import Any, Callable, Dict, Iterable, Iterator, Sequence
+
+
+class Traversable:
+
+    __slots__ = ()
+
+    @property
+    @abstractmethod
+    def __children__(self) -> Sequence[Traversable]:
+        ...
+
+
+class Graph(Dict[Traversable, Sequence[Traversable]]):
+    def __init__(self, mapping=(), /, **kwargs):
+        if isinstance(mapping, Traversable):
+            mapping = self.bfs(mapping)
+        super().__init__(mapping, **kwargs)
+
+    @classmethod
+    def bfs(cls, root: Traversable) -> Graph:
```

Comment:
Can we call this `from_bfs`/`from_dfs`? It feels like `dfs` and `bfs` would be instance methods if anything.

Reply:
Updated.